File: matlab/task1_6.m
% 1.6. Limiarização aplicada sobre Y, com limiar m e duas opções: a) m
% escolhido pelo usuário; b) m = média de valores da banda Y;
function task1_6()
    assetsDir = ['..' filesep 'assets' filesep];
    filePath = strcat(assetsDir, 'lenna.png');
    originalImage = imread(filePath);
    
    FIXED_M = 0.2;
    yiqImage = rgb2ntsc(originalImage);
    grayScaleImage = yiqImage(:,:,1);
         
    % Segmented Y channel with fixed M
    fixedMOutput = binarySegmention(grayScaleImage, FIXED_M);
    fixedMImage = yiqImage;
    fixedMImage(:,:,1) = fixedMOutput;
    fixedMBack2RGB = ntsc2rgb(fixedMImage);
        
    % Segmented Y channel with M = mean(Y)
    meanM = mean(grayScaleImage(:));
    meanMOutput = binarySegmention(grayScaleImage, meanM);
    meanMImage = yiqImage;
    meanMImage(:,:,1) = fixedMOutput;
    meanMBack2RGB = ntsc2rgb(meanMImage);

    % Ploting
    fig=figure();        
    subplot(1,3,1), imshow(originalImage), title('Original image')    
    subplot(1,3,2), imshow(fixedMOutput), title(strcat('Segmented Y with Fixed M = ',num2str(FIXED_M)))        
    subplot(1,3,3), imshow(fixedMBack2RGB), title('Back to RGB')
       
    fig2=figure();        
    subplot(1,3,1), imshow(originalImage), title('Original image')        
    subplot(1,3,2), imshow(meanMOutput), title(strcat('Segmented Y with M = mean(Y) = ',num2str(meanM)))
    subplot(1,3,3), imshow(meanMBack2RGB), title('Back to RGB')
    
    saveas(fig,strcat('output',filesep,mfilename,'_1.png'));
    saveas(fig2,strcat('output',filesep,mfilename,'_2.png'));
end

function output = binarySegmention(input, m)
    output = (input>m);            
end

File: python/task1_6.py
# coding=UTF-8
# 1.6. Limiarização aplicada sobre Y, com limiar m e duas opções: a) m
# escolhido pelo usuáio; b) m = média de valores da banda Y;

import numpy as np
import utils
import color


def main():
    image = utils.load_image('lenna.png')

    yiq_image = color.rgb2yiq(image)
    grayscale_image = color.yiq2rgb(image)[:, :, 0]

    threshold_value = 255 * 0.2
    mean_value = np.mean(grayscale_image)

    threshold_user_image = _segment(grayscale_image, threshold_value)
    original_threshold_user_image = np.copy(yiq_image)
    original_threshold_user_image[:, :, 0] = threshold_user_image
    original_threshold_user_image = color.yiq2rgb(original_threshold_user_image)

    threshold_mean_image = _segment(grayscale_image, mean_value)
    original_threshold_mean_image = np.copy(yiq_image)
    original_threshold_mean_image[:, :, 0] = threshold_mean_image
    original_threshold_mean_image = color.yiq2rgb(original_threshold_mean_image)

    utils.display_single_image('Original Image', image)
    utils.display_single_image('YIQ Image', yiq_image)
    utils.display_single_image('Y Channel', grayscale_image)

    utils.display_single_image('Y Threshold (User ' + str(threshold_value) + ')', threshold_user_image)
    utils.display_single_image('Back to Original (User ' + str(threshold_value) + ')', original_threshold_user_image)

    utils.display_single_image('Y Threshold (Mean ' + str(mean_value) + ')', threshold_mean_image)
    utils.display_single_image('Back to Original (Mean ' + str(mean_value) + ')', original_threshold_mean_image)

    utils.wait_key_and_destroy_windows()


def _segment(image, m):
    output = (image > m) * 255
    return output

if __name__ == "__main__":
    main()